### PR TITLE
Sprint 1 – Add Benchmarking_and_Experimentation/config updates

### PR DESCRIPTION
**Pull request by 223089688**

### CODE REVIEW

File: src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py [hunk 1]
```diff
@@ -1,246 +1,151 @@
...
+    backbone = backbone_class(
+        include_top=False,
+        weights="imagenet",
+        input_shape=(h, w, c),
+        pooling="avg"
     )
+    backbone.trainable = True
+    y = backbone(x, training=False)
+
+    reg = tf.keras.regularizers.l2(l2_coefficient) if l2_regularization else None
```
Comment (newbee1905) [inline src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py:63]: I think using BatchNormalization would make more sense here rather than normal regularization
or is this part of your architecture?

File: src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py [hunk 2]
```diff
@@ -1,246 +1,151 @@
...
+    sample_batch = next(iter(train_ds.take(1)))
+    if len(sample_batch[1].shape) == 1:
+        loss_fn = "sparse_categorical_crossentropy"
+
+    lr = float(MODELS.get(model_name, {}).get("learning_rate", 1e-4))
     model.compile(
-        optimizer=optimizer,
-        loss=tf.keras.losses.CategoricalCrossentropy(from_logits=True),
-        metrics=['accuracy']
+        optimizer=tf.keras.optimizers.Adam(learning_rate=lr),
```
Comment (newbee1905) [inline src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py:125]: If you have extra time, make optimizer to be an optional configuration as well, so people can test out between SGD, Adam, AdamW, etc.

File: src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py [hunk 3]
```diff
@@ -1,246 +1,151 @@
...
-    # Define callbacks based on the notebook example
-    log_dir = os.path.join("tensorboard_logs", model_name + "_" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
-    tensorboard_callback = TensorBoard(log_dir=log_dir, histogram_freq=1)
+    callbacks = []
+    if use_callbacks:
+        callbacks = [
+            tf.keras.callbacks.EarlyStopping(monitor="val_loss", patience=3, restore_best_weights=True),
+            tf.keras.callbacks.ReduceLROnPlateau(monitor="val_loss", factor=0.5, patience=2),
+            tf.keras.callbacks.TensorBoard(log_dir=os.path.join(SC.get("OUTPUT_DIRECTORY", "."), "logs"))
+        ]
```
Comment (newbee1905) [inline src/Prototypes/engine/Benchmarking_and_Experimentation/utils/optimised_engine_pipeline.py:137]: similar to optimizer, if you have extra time, please also add in option to change scheduler, for scheduler, having no scheduler would also an option as well

**Review by newbee1905 — APPROVED** [review]
Well done
